WIP: Implement LDAP Support

Polynomdivision opened the pull request:
This WIP PR (fixed #53) adds support for optionally adding LDAP into the login and registration procedure by looking the user up in
the LDAP directory to decide whether to proceed with the login/registration.

While the login code is tested and working, the registration code is untested.

A question that was brought up in #53 is how to deal with a user that is logged in but whose access in LDAP is then
revoked. I am unsure how other applications do this. My guess is that they wait until the session of the user expires and the user
has to reauthenticate. But since tokens appear to live for ~30 days this might not work. The most elegant solution would probably be to cache the result of the user lookup for some time and use this on every authenticated API path (as suggested in #53).

tasn commented:
Hey,

Thanks for the PR! I haven't reviewed it properly, but my biggest concern is the hardcoded dep on LDAP. I wonder if there's a way to add it that is less involved.

Also, the changes shouldn't be touching `django_etebase` directory at all. Another approach would be to maybe move the `ldap.py` file under `myauth` (as it's an auth backend), and also have a `create_user` there. This `create_user` can then be used by setting `ETEBASE_CREATE_USER_FUNC = 'myauth.ldap.create_user'` in the `settings.py` file (for those who want to enable it). This will make it much more contained and essentially only called if enabled.
Another thing is: have the settings under `LDAP_SETTINGS` or something like that. No need to take over the `ETEBASE` namespace which is used for django_etebase.

django_etebase is only for etebase protocol specific stuff.

Polynomdivision commented:
As I've said, I never really worked with Django so you have to excuse my mistakes. I didn't know that the user is supposed to work
with the settings.py (via etebase_server_settings.py). I always just went with the config file.

I have a new version that addresses your issues by moving ldap.py into myauth/ and implementing there a Django permission and a new user_create function that the user can include. Again, logging in I have briefly tested but the registration is untested.

tasn commented:
Looks much better, maybe also set `ETEBASE_CREATE_USER_FUNC` in https://github.com/etesync/server/pull/64/files#diff-e09efb04d7681243823b754c3baa8b800bdc68f3fc7799d715e0db56d7d543f4R167 if ldap is indeed set.
Dunno if it's the right thing, but potentially.
I'd also put the permissions in the same file and then add the permission to the permissions array if ldap is enabled.

Anyhow, looks good overall! Keep on testing it and once complete we can merge it in!

Polynomdivision commented:
Automatically setting the create_user function and adding the permission is no problem. But I don't understand what you mean by putting the permissions in the same file.

tasn commented:
Haven't properly reviewed it yet, but just a note for myself/putting it out there: maybe it'd be a good idea to integrate something like https://www.keycloak.org/ that can take care of all of this and more.

Polynomdivision commented:
I have just deployed my branch on my own server to see if anything breaks. I'll let you know in a week or so if I noticed any major hiccups. So far, the login works on my real LDAP server.

tasn commented:
I'm so sorry, I still haven't managed to review this. :( 
Additionally, we just added Black auto-formatting for strict code validation (needed now that there are more and more external contributions), could you please fix the conflicts in your PR + run `Black` to autoformat your changes? Thanks!

Polynomdivision commented:
> I'm so sorry, I still haven't managed to review this. :(

No worries. Take your time. This PR can wait :smile:
> [...] could you please fix the conflicts in your PR + run `Black` to autoformat your changes? Thanks!

Done. Though I just noticed that I also would like to implement another LDAP setting to set how long a cached LDAP entry
should be valid. Putting this one on my TODO list.

tasn commented:
How is it going so far? Is it working well for you on your server?

I really wonder about the keycloack integration as an alternative option that also supports caldav along many other things.

Polynomdivision commented:
> How is it going so far? Is it working well for you on your server?

My branch has been running on my server for about a week now and I have not had any issues related to login. Registration is still untested, probably because I once wanted to test it, but after pressing "register" I did not see a HTTP request in the console, which spooked me a little :sweat_smile:

I still want to make the LDAP cache TTL configurable, but I have not found time.

> I really wonder about the keycloack integration [...]

Sadly I cannot comment on this one as I have never worked with - or even looked at - keycloak. But based on the phrasing on their homepage
saying that it allows application developers to easily add authentication without caring about users **and** have a connection to an LDAP directory, it sounds nice.

poVoq commented:
Keycloak is AFAIK a huge Java codebase and not something that is easily deployed. Might be really overkill.

But there is this new WIP project that is small and also written in Python that could be a better companion for Etebase: https://gitlab.com/yaal/canaille

mdmdmdmdmd commented:
I noticed at https://github.com/etesync/server/pull/64/files#diff-1fb90ee86b332fb1f87f3867782778fd081f871065f0c3dd55664a0acf9ed516R94 that 'User' has never been defined before its usage there. Is that intentional? Seems like you would need to do: 'from django.contrib.auth import get_user_model' and 'User = get_user_model()'. Which I tried, but even then i get an error on sign up saying "User() got an unexpected keyword argument 'view'"

Polynomdivision commented:
@mdmdmdmdmd Oh yeah, I must've missed that. That was not intentional. I basically just took the `create_user` function from
django_etebase.utils but added my own check before actually creating the user, so I have no idea why it doesn't work. I will try to fix this but I currently have no time to work on this. It will go into my queue though ;)

Polynomdivision commented:
@mdmdmdmdmd Sorry it took so long. I found time again and fixed the issue you had. Turns out that I forgot to copy something over from the original create_user function.

However, I couldn't really test it as some complications occurred. But I will update this PR once I have solved those.

Polynomdivision commented:
Okay. I now have fixed and tested user creation. It should work now.

tasn commented on etebase_server/settings.py:
```diff
@@ -130,6 +130,11 @@
 MEDIA_ROOT = os.environ.get("DJANGO_MEDIA_ROOT", os.path.join(BASE_DIR, "media"))
 MEDIA_URL = "/user-media/"
 
+ETEBASE_API_PERMISSIONS = ["rest_framework.permissions.IsAuthenticated"]
```
These two settings are wrong, we removed them in master. Are they even needed for you, or just a merge mistake?

Polynomdivision replied:
Yes, they were a merge mistake.

tasn commented on etebase_server/settings.py:
```diff
@@ -138,6 +143,8 @@
     "/etc/etebase-server/etebase-server.ini",
 ]
 
+ETEBASE_CREATE_USER_FUNC = "django_etebase.utils.create_user_blocked"
```
I guess this is also a merge issue because it was just moved to a different place in your commit.

tasn reviewed (commented):
Made a few comments.

Polynomdivision commented:
I would say that the PR now works as intended with master.

Polynomdivision commented:
As I was very busy the last months, I am happy to say that I finally found time to work on this PR again. I just deployed this version to my server and will say whether I encountered any issues in about two weeks.

If there are any other issues with this PR, I would be happy to fix them.

tasn commented on etebase_server/settings.py:
```diff
@@ -163,7 +163,17 @@
     if "database" in config:
         DATABASES = {"default": {x.upper(): y for x, y in config.items("database")}}
 
-ETEBASE_CREATE_USER_FUNC = "django_etebase.utils.create_user_blocked"
```
This line shouldn't be removed, shouldn't it? Or is it a merge issue?

Polynomdivision replied:
It should be fixed now.

tasn commented:
Great! It conflicts with master at the moment (we changed a few things), and I made a small comment, though otherwise I think it's fine to put it in.

DoM1niC commented:
I use this wonderful PR but how I setup the Config exactly 

like this ? 

```
[ldap]
ldap_server = ldap://SERVER:389
ldap_search_base = "dc=home"
ldap_filter = "(o=true)"
```

???

Polynomdivision commented:
@DoM1niC It's almost correct. My approach, if I remember correctly, only works with a Bind DN, e.g. a user to authenticate as that authenticates the other users.

So you would also need to add:

```
bind_dn = <DN to bind as>
bind_pw = <Password>
```

Also, the options are called `server`, `search_base` and `filter`. The `ldap_` prefix is not correct.

DoM1niC commented:
I use anonymous bind btw

DoM1niC commented:
> (&(uid=%%s)(o=true))

ok I understand so far but still show `401 Unauthorized`

Polynomdivision commented:
@DoM1niC Are you using the password of the LDAP user or the Etebase password? It's been a long time since I last looked at the code, but I remember that my LDAP code doesn't check the password, only if the user exists. The password authentication process that follows in the regular Etesync code.

tasn commented:
Do we feel like this PR is ready? Or is it at risk of failure?

Polynomdivision commented:
I mean, I have been running this code pretty much since the last comment that I'm using this PR without the EteSync app ever telling me that it couldn't sync.

Polynomdivision commented:
I have now added support for reading the LDAP password from a file for running it on systems like NixOS

Xiretza commented:
What's the status on this? @PapaTutuWawa looks like it needs another rebase (and a cleanup of the existing rebase errors), @tasn would you be ready to merge it once it's rebased?

tasn commented:
Yeah @Xiretza, I think it's time.

PapaTutuWawa commented:
I have fixed the merge conflict.

Xiretza commented:
Can you squash those commits down to one? There's really no reason to have 21 commit, including 3 merges, for these relatively small changes.

Xiretza commented on myauth/ldap.py:
```diff
@@ -0,0 +1,111 @@
+from django.utils import timezone
+from django.conf import settings
+from django.core.exceptions import PermissionDenied as DjangoPermissionDenied
+from django_etebase.utils import CallbackContext
```
These (`django_etebase`, `myauth`, `etebase_fastapi`) all need to be updated according to #140.

Xiretza commented on etebase_server/settings.py:
```diff
@@ -167,8 +169,20 @@
+        LDAP_CACHE_TTL = ldap.get("cache_ttl", "")
+
+        # Configure EteBase to use LDAP
+        ETEBASE_CREATE_USER_FUNC = "myauth.ldap.create_user"
+        ETEBASE_API_PERMISSIONS_READ = ["myauth.ldap.is_user_in_ldap"]
```
Ditto.

Xiretza commented on etebase_server/settings.py:
```diff
@@ -140,6 +140,8 @@
     "/etc/etebase-server/etebase-server.ini",
 ]
 
+ETEBASE_CREATE_USER_FUNC = "django_etebase.utils.create_user_blocked"
```
This rebase/merge problem is still here.

Xiretza commented on etebase_server/settings.py:
```diff
@@ -167,8 +169,20 @@
+        
+    if "ldap" in config:
+        ldap = config["ldap"]
+        LDAP_SERVER = ldap.get("server", "")
+        LDAP_SEARCH_BASE = ldap.get("search_base", "")
+        LDAP_FILTER = ldap.get("filter", "")
+        LDAP_BIND_DN = ldap.get("bind_dn", "")
+        LDAP_BIND_PW = ldap.get("bind_pw", "")
+        LDAP_BIND_PW_FILE = ldap.get("bind_pw_file", "")
+        LDAP_CACHE_TTL = ldap.get("cache_ttl", "")
```
Should these really have empty strings as defaults? Wouldn't it make more sense to error if they aren't specified (or for `bind_pw`/`bind_pw_file`, if neither of them is specified)?

PapaTutuWawa commented:
Okay, everything should be rebased and adjusted for PR#140 now. I have also tested it on my server, just to make sure. I also added an error if either no BindDN or a password is specified.

Xiretza commented:
Thanks for the quick work, looks good to me now!

tasn commented:
Happy to have it in! 🎉 🎉 🎉 

Apologies to @PapaTutuWawa for taking so long to get it in and thanks again for the contribution!
Also thanks to @Xiretza for reviving the conversation!

@Xiretza, what do you think, should we make a new release? Are we confident with the state of things?

Xiretza commented:
I have to admit I only got around to actually setting up a production server a couple days ago, but it's been running off master without problems so far. The gnarliest issues would probably be some kind of migration problem, but I don't think anything related to the database has changed since 0.8.3.

tasn commented:
Yeah, I don't think so either. OK, I'll take a look at making a new release.
